WAL failover: deemphasize remote log sinks (#18884)
cockroachdb/docs: src/current/v24.1/cockroach-start.md

> [Writing log files to local disk]({% link {{ page.version.version }}/configure-logs.md %}#output-to-files) using the default configuration can lead to cluster instability in the event of a [disk stall]({% link {{ page.version.version }}/cluster-setup-troubleshooting.md %}#disk-stalls). So It's not enough to failover your WAL writes to another disk, you must also write your log files in such a way that the forward progress of your cluster is not stalled due to disk unavailability.

Author: Fixed

Therefore, if you enable WAL failover and log to local disks, you must also update your [logging]({% link {{page.version.version}}/logging-overview.md %}) configuration as follows:

> 1. Disable [audit logging]({% link {{ page.version.version }}/sql-audit-logging.md %}). File-based audit logging and the WAL failover feature cannot coexist. File-based audit logging provides guarantees that every log message makes it to disk, otherwise CockroachDB needs to shut down. Because of this, resuming operations in the face of disk unavailability is not compatible with audit logging.

Reviewer: Some optional suggestions.

```suggestion
1. Disable [audit logging]({% link {{ page.version.version }}/sql-audit-logging.md %}). File-based audit logging cannot coexist with the WAL failover feature. File-based audit logging guarantees that every log message makes it to disk, or CockroachDB must be shut down. For this reason, resuming operations in the face of disk unavailability is not compatible with audit logging.
```

Author: Fixed

> 1. Enable asynchronous buffering of [`file-groups` log sinks]({% link {{ page.version.version }}/configure-logs.md %}#output-to-files) using the `buffering` configuration option. The `buffering` configuration can be applied to [`file-defaults`]({% link {{ page.version.version }}/configure-logs.md %}#configure-logging-defaults) or individual `file-groups` as needed. Note that enabling asynchronous buffering of `file-groups` log sinks is in [preview]({% link {{page.version.version}}/cockroachdb-feature-availability.md %}#features-in-preview).

Reviewer: This confused me because the `buffering` option isn't documented in the table at `configure-logs.html#output-to-files`. Do you know if we have a separate ticket to cover that? It's also not listed in the linked 'preview features' doc as far as I can tell.

Author: ya it's a bit confusing, since this feature is not GA we didn't yet doc the log buffering stuff

from my other comment:

> totes agreed, right now it's experimental per https://cockroachlabs.atlassian.net/browse/DOC-9925 bcs it was paired with the WAL failover stuff

> but we have https://cockroachlabs.atlassian.net/browse/DOC-10240 which will add it when it becomes GA

> 1. Set `max-staleness: 1s` and `flush-trigger-size: 256KiB`.

Reviewer: Not sure why `max-buffer-size` isn't mentioned here but is in the example. Is it less important?

EDIT: Now that I've read all the steps, I feel like this can be cut entirely, since it's all in the snippet in the final step.

```suggestion
1. Set the `buffering` configuration to `max-staleness: 1s`, `flush-trigger-size: 256KiB`, and `max-buffer-size: 50MiB`.
```

Author: not gonna take this edit bc idk if it's technically necessary and i don't wanna touch technical stuff that's already had a SME review

> 1. When `buffering` is enabled, `buffered-writes` must be explicitly disabled as shown below. This is necessary because `buffered-writes` does not provide true asynchronous disk access, but rather a small buffer. If the small buffer fills up, it can cause internal routines performing logging operations to hang. This in turn will cause internal routines doing other important work to hang, potentially affecting cluster stability.

Reviewer: ```suggestion
1. When `buffering` is enabled, `buffered-writes` must be explicitly disabled as shown in the following example. This is necessary because `buffered-writes` does not provide true asynchronous disk access, but rather a small buffer. If the small buffer fills up, it can cause internal routines performing logging operations to hang. This will in turn cause internal routines doing other important work to hang, potentially affecting cluster stability.
```

Author: Fixed
